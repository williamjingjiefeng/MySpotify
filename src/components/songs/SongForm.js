import React from "react";
import PropTypes from "prop-types";
import TextInput from "../common/TextInput";
import SelectInput from "../common/SelectInput";

const SongForm = ({
  song,
  singers,
  albums,
  onSave,
  onChange,
  saving = false,
  errors = {},
}) => {
  return (
    <form onSubmit={onSave}>
      <h2>{song.id ? "Edit" : "Add"} Song</h2>
      {errors.onSave && (
        <div className="alert alert-danger" role="alert">
          {errors.onSave}
        </div>
      )}
      <TextInput
        name="title"
        label="Title"
        value={song.title}
        onChange={onChange}
        error={errors.title}
      />

      <SelectInput
        name="singerId"
        label="Singer"
        value={song.singerId || ""}
        defaultOption="Select Singer"
        options={singers.map((singer) => ({
          value: singer.id,
          text: singer.name,
        }))}
        onChange={onChange}
        error={errors.singer}
      />

      <TextInput
        name="youtubeId"
        label="Youtube ID"
        value={song.youtubeId}
        onChange={onChange}
        error={errors.youtubeId}
      />

      <SelectInput
        name="albumId"
        label="Album"
        value={song.albumId || ""}
        defaultOption="Select Album"
        options={albums.map((album) => ({
          value: album.id,
          text: album.name,
        }))}
        onChange={onChange}
        error={errors.album}
      />

      <button type="submit" disabled={saving} className="btn btn-primary">
        {saving ? "Saving..." : "Save"}
      </button>
    </form>
  );
};

SongForm.propTypes = {
  singers: PropTypes.array.isRequired,
  albums: PropTypes.array.isRequired,
  song: PropTypes.object.isRequired,
  errors: PropTypes.object,
  onSave: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  saving: PropTypes.bool,
};

export default SongForm;
