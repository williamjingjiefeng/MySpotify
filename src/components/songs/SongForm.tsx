import React from "react";
import TextInput from "../common/TextInput";
import SelectInput from "../common/SelectInput";
import { ISong, ISinger, IAlbum, IErrors } from "../../redux/dispatch/Music/PreferenceState";

interface ISongForm {
    song: ISong;
    singers: ISinger[];
    albums: IAlbum[];
    onSave: (event: React.FormEvent<HTMLFormElement>) => void;
    onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
    saving: boolean;
    errors: IErrors;
}

const SongForm = ({
    song,
    singers,
    albums,
    onSave,
    onChange,
    saving = false,
    errors,
}: ISongForm) => {
    return (
        <form onSubmit={onSave}>
            <h2>{song.id ? "Edit" : "Add"} Song</h2>
            {errors.onSave && (
                <div className="alert alert-danger" role="alert">
                    {errors.onSave}
                </div>
            )}
            <TextInput
                name="title"
                label="Title"
                value={song.title}
                onChange={onChange}
                error={errors.title}
            />

            <SelectInput
                name="singerId"
                label="Singer"
                value={song.singerId || ""}
                defaultOption="Select Singer"
                options={singers.map((singer) => ({
                    value: singer.id,
                    text: singer.name,
                }))}
                onChange={onChange}
                error={errors.singer}
            />

            <TextInput
                name="youtubeId"
                label="Youtube ID"
                value={song.youtubeId}
                onChange={onChange}
                error={errors.youtubeId}
            />

            <SelectInput
                name="albumId"
                label="Album"
                value={song.albumId || ""}
                defaultOption="Select Album"
                options={albums.map((album) => ({
                    value: album.id,
                    text: album.name,
                }))}
                onChange={onChange}
                error={errors.album}
            />

            <button type="submit" disabled={saving} className="btn btn-primary">
                {saving ? "Saving..." : "Save"}
            </button>
        </form>
    );
};

export default SongForm;
