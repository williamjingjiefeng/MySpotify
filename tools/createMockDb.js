/* eslint-disable no-console */
const fs = require("fs");
const path = require("path");
const mockMusic = require("./mockMusics");

const { songs, singers, albums } = mockMusic;
const data = JSON.stringify({ songs, singers, albums });
const filepath = path.join(__dirname, "db.json");

fs.writeFile(filepath, data, function (err) {
  err ? console.log(err) : console.log("Mock DB created.");
});
