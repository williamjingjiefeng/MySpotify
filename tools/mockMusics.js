const songs = [
  {
    id: 1,
    title: "Scarborough Fair",
    youtubeId: "oJLNCTdJKu4",
    singerId: 3,
    albumId: 1,
  },
  {
    id: 2,
    title: "Monsters",
    youtubeId: "xHWIhCUnfOE",
    singerId: 6,
    albumId: 2,
  },
  {
    id: 3,
    title: " Let Me Down Slowly",
    youtubeId: "50VNCymT-Cs",
    singerId: 5,
    albumId: 3,
  },
  {
    id: 4,
    title: "Where Do I Begin",
    youtubeId: "kUrxm4z9I-c",
    singerId: 4,
    albumId: 4,
  },
  {
    id: 5,
    title: "You Raise Me Up",
    youtubeId: "aJxrX42WcjQ",
    singerId: 2,
    albumId: 5,
  },
  {
    id: 6,
    title: "Ypu're The Place",
    youtubeId: "vv_bzipW_m0",
    singerId: 2,
    albumId: 5,
  },
  {
    id: 7,
    title: "Se√±orita",
    youtubeId: "O6gdUM3ffRQ",
    singerId: 7,
    albumId: 6,
  },
  {
    id: 8,
    title: "Don't Cry For Me Argentina",
    youtubeId: "KD_1Z8iUDho",
    singerId: 8,
    albumId: 7,
  },
];

const singers = [
  { id: 1, name: "Katy Perry" },
  { id: 2, name: "Josh Groban" },
  { id: 3, name: "Sarah Brigtman" },
  { id: 4, name: "Andy Williams" },
  { id: 5, name: "Alec Benjamin" },
  { id: 6, name: "Katie Sky" },
  { id: 7, name: "Shawn Mendes & Camila Cabello" },
  { id: 8, name: "Madonna" },
];

const albums = [
  { id: 1, singerId: 3, name: "La Luna" },
  { id: 2, singerId: 6, name: "After Hours" },
  { id: 3, singerId: 5, name: "Narrated For You" },
  { id: 4, singerId: 4, name: "LOVE STORY 1970" },
  { id: 5, singerId: 2, name: "Closer" },
  { id: 6, singerId: 7, name: "Shawn Mendes" },
  { id: 7, singerId: 8, name: "Evita" },
];

const newSong = {
  id: null,
  title: "",
  singerId: null,
  youtubeId: "",
  albumId: null,
};

// Using CommonJS style export so we can consume via Node (without using Babel-node)
module.exports = {
  newSong: newSong,
  songs: songs,
  singers: singers,
  albums: albums,
};
