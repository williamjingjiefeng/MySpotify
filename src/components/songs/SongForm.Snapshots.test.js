import React from "react";
import SongForm from "./SongForm";
import renderer from "react-test-renderer";
import { songs, singers, albums } from "../../../tools/mockMusics";

it("sets submit button label 'Saving...' when saving is true", () => {
  const tree = renderer.create(
    <SongForm
      song={songs[0]}
      singers={singers}
      albums={albums}
      onSave={jest.fn()}
      onChange={jest.fn()}
      saving
    />
  );

  expect(tree).toMatchSnapshot();
});

it("sets submit button label 'Save' when saving is false", () => {
  const tree = renderer.create(
    <SongForm
      song={songs[0]}
      singers={singers}
      albums={albums}
      onSave={jest.fn()}
      onChange={jest.fn()}
      saving={false}
    />
  );

  expect(tree).toMatchSnapshot();
});
