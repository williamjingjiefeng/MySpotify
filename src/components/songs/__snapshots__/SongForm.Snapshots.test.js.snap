// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sets submit button label 'Save' when saving is false 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <h2>
    Edit
     Song
  </h2>
  <div
    className="form-group"
  >
    <label
      htmlFor="title"
    >
      Title
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="title"
        onChange={[MockFunction]}
        type="text"
        value="Scarborough Fair"
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="singerId"
    >
      Singer
    </label>
    <div
      className="field"
    >
      <select
        className="form-control"
        name="singerId"
        onChange={[MockFunction]}
        value={3}
      >
        <option
          value=""
        >
          Select Singer
        </option>
        <option
          value={1}
        >
          Katy Perry
        </option>
        <option
          value={2}
        >
          Josh Groban
        </option>
        <option
          value={3}
        >
          Sarah Brigtman
        </option>
        <option
          value={4}
        >
          Andy Williams
        </option>
        <option
          value={5}
        >
          Alec Benjamin
        </option>
        <option
          value={6}
        >
          Katie Sky
        </option>
        <option
          value={7}
        >
          Shawn Mendes & Camila Cabello
        </option>
        <option
          value={8}
        >
          Madonna
        </option>
      </select>
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="youtubeId"
    >
      Youtube ID
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="youtubeId"
        onChange={[MockFunction]}
        type="text"
        value="oJLNCTdJKu4"
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="albumId"
    >
      Album
    </label>
    <div
      className="field"
    >
      <select
        className="form-control"
        name="albumId"
        onChange={[MockFunction]}
        value={1}
      >
        <option
          value=""
        >
          Select Album
        </option>
        <option
          value={1}
        >
          La Luna
        </option>
        <option
          value={2}
        >
          After Hours
        </option>
        <option
          value={3}
        >
          Narrated For You
        </option>
        <option
          value={4}
        >
          LOVE STORY 1970
        </option>
        <option
          value={5}
        >
          Closer
        </option>
        <option
          value={6}
        >
          Shawn Mendes
        </option>
        <option
          value={7}
        >
          Evita
        </option>
      </select>
    </div>
  </div>
  <button
    className="btn btn-primary"
    disabled={false}
    type="submit"
  >
    Save
  </button>
</form>
`;

exports[`sets submit button label 'Saving...' when saving is true 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <h2>
    Edit
     Song
  </h2>
  <div
    className="form-group"
  >
    <label
      htmlFor="title"
    >
      Title
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="title"
        onChange={[MockFunction]}
        type="text"
        value="Scarborough Fair"
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="singerId"
    >
      Singer
    </label>
    <div
      className="field"
    >
      <select
        className="form-control"
        name="singerId"
        onChange={[MockFunction]}
        value={3}
      >
        <option
          value=""
        >
          Select Singer
        </option>
        <option
          value={1}
        >
          Katy Perry
        </option>
        <option
          value={2}
        >
          Josh Groban
        </option>
        <option
          value={3}
        >
          Sarah Brigtman
        </option>
        <option
          value={4}
        >
          Andy Williams
        </option>
        <option
          value={5}
        >
          Alec Benjamin
        </option>
        <option
          value={6}
        >
          Katie Sky
        </option>
        <option
          value={7}
        >
          Shawn Mendes & Camila Cabello
        </option>
        <option
          value={8}
        >
          Madonna
        </option>
      </select>
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="youtubeId"
    >
      Youtube ID
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="youtubeId"
        onChange={[MockFunction]}
        type="text"
        value="oJLNCTdJKu4"
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="albumId"
    >
      Album
    </label>
    <div
      className="field"
    >
      <select
        className="form-control"
        name="albumId"
        onChange={[MockFunction]}
        value={1}
      >
        <option
          value=""
        >
          Select Album
        </option>
        <option
          value={1}
        >
          La Luna
        </option>
        <option
          value={2}
        >
          After Hours
        </option>
        <option
          value={3}
        >
          Narrated For You
        </option>
        <option
          value={4}
        >
          LOVE STORY 1970
        </option>
        <option
          value={5}
        >
          Closer
        </option>
        <option
          value={6}
        >
          Shawn Mendes
        </option>
        <option
          value={7}
        >
          Evita
        </option>
      </select>
    </div>
  </div>
  <button
    className="btn btn-primary"
    disabled={true}
    type="submit"
  >
    Saving...
  </button>
</form>
`;
